mock-api/msw/handlers.ts

```diff
@@ -486,6 +486,33 @@ export const handlers = makeHandlers({
       time_run_state_updated: new Date().toISOString(),
     }
 
+    body.external_ips?.forEach((ip) => {
+      if (ip.type === 'floating') {
+        const floatingIp = lookup.floatingIp({
+          project: project.id,
+          floatingIp: ip.floating_ip,
+        })
+        if (floatingIp.instance_id) {
+          throw 'floating IP cannot be attached to one instance while still attached to another'
+        }
+        floatingIp.instance_id = instanceId
+      } else if (ip.type === 'ephemeral') {
+        const poolId = db.ipPools.find((pool) => pool.name === ip.pool)?.id
+        const range = db.ipPoolRanges.find((range) => range.ip_pool_id === poolId)?.range
```

Comment:
If the pool has no ranges you should 400. I think that's what the API does (it may 404 instead of 400, not sure).

I think it would make sense to wrap these three lines in a util function `getIpFromPool` that takes the pool `NameOrId` just like the lookup function. The functionality won't be right but we can use the function everywhere currently allocate an IP (don't make that change now unless it's like two spots) and then we can get the logic right later. One interesting question is whether this function should 400 if there are no ranges or it should just return `null` and the calling code decides what kind of error that is. Different call sites might want to handle it differently? I don't know. There's also the problem of distinguishing between not getting an IP because the pool doesn't exist, not getting an IP because there are no ranges, and not getting an IP because the pool is exhausted. Returning null from the function wouldn't be sufficient for that. I'm also not 100% sure what the 4xx status code from the API is in each of those cases.